### ISSUE TITLE
Conditional “X-Prompt” in Angular Schematics

### ISSUE THREAD

**Issue by Ankit-sagar32** [angular/angular-cli#18578]
It seems feature to add conditional x-prompts is missing in angular schematics.

**Requirement** : To display conditional prompts, based on the previous selection ( "Analytics", in this case ).

**Problem** : For any selection, it is still displaying the prompt to add Google Analytics's Tracking Id even if i select AppDynamics or none.

**References** : [Applying Subschemas Conditionally](https://json-schema.org/understanding-json-schema/reference/conditionals.html)
                           [Angular Schematics](https://angular.io/guide/schematics-authoring)

**Note**: For Simplicity, i have removed "allOf","anyOf" etc. and code for any other analytics tool, still it was not working.

**Code** (Schema.json):
```
{
  "$schema": "http://json-schema.org/schema",
  "id": "MyDemoSchema",
  "title": "My Demo Schematics Schema",
  "type": "object",
  "properties": {
    "analytics": {
      "enum": [
        "none",
        "GoogleAnalytics",
        "AppDynamics"
      ],
      "x-prompt": "Which Analytics would you like to use?"
    }
  },
  "if": {
    "properties": {
      "analytics": {
        "const": "GoogleAnalytics"  
      }
    }
  },
  "then": {
    "properties": {
      "GoogleAnalytics": {
        "type": "string",
        "x-prompt": "Enter Tracking ID?"
      }
    }
  }
}
```

**Comment by Stupidism** [comment]
Any progress or workaround?

**Comment by kroeder** [comment]
I have the same goal but a different proposal. Instead of increasing the basic schema.json feature-set, I'd rather use a second function-driven API. I believe adding a condition API to a json file will increase the complexity tremendously.

Here's a proposal. It certainly contains lots of flaws and is meant as a start to discuss further possible approaches.

**collection.json**
```json
{
  "$schema": "../../../node_modules/@angular-devkit/schematics/collection-schema.json",
    "dynamic-generate-props": {
        "description": "Dynamic inputs",
        "factory": "./dynamic/index#dynamicSchematic",
        "schema": "./dynamic/schema#conditionalInputs"
    }
  }
}
```

**dynamic/schema.ts#conditionalInputs**
```typescript
/**
 * User code
 */
interface MyInput {
    name: string;
    path: string;
    needsFooter: boolean;
    footerColor?: FooterColor;
}

type FooterColor = 'blue' | 'green' | 'red';

export function conditionalInputs() {
    return schematicsSchema<MyInput>({
        name: () => createInput({
            type: InputTypes.STRING,
            xPrompt: "Name of your component"
        }),
        path: () => executionPath(),
        needsFooter: () => createInput({
            type: InputTypes.BOOLEAN,
            xPrompt: "Do you need a footer?"
        }),
        footerColor: (prevValues) => prevValues.needsFooter 
            ? noop()
            : createInput({
                type: InputTypes.ENUM,
                enum: [
                    "blue",
                    "green",
                    "red"
                ],
                xPrompt: "Footer color"
            })
    })
}
```

**Angular framework (pseudo code)**
```typescript
enum InputTypes { STRING, BOOLEAN, NUMBER, ENUM }
type InputFunction<TInput> = { [key: string]: (prevValues?: TInput) => any; }

function schematicsSchema<TInput>(userInputs: InputFunction<TInput>) {
    return {};
}

function createInput(options: any) {}

function executionPath() {}

function noop() {}
```

Having an interface in the user code is intentional. Right now, if I am writing a schematic, the input properties are not typed and I need to manage two lists of inputs in parallel (schema.json inputs / schematic factory code input type)

Any thoughts on this? Or maybe disagreement to a TS API? 🙂

**Comment by Jordan-Hall** [comment]
> I have the same goal but a different proposal. Instead of increasing the basic schema.json feature-set.

This wouldn't exactly be increasing the schema.json rather than implementing draft-7 of Json schema.

As for your typing solution https://ajv.js.org/guide/schema-language.html#json-type-definition I imagine that would be good enough. 

Angular currently uses AJV for validation and running the registry. Looking through the code I cant see it being to difficult to add support for a typed implementation.

With that said I think to add xprompt support its changing the code around this section https://github.com/angular/angular-cli/blob/fafb9ddd589e02643b240da7906589737077e325/packages/angular_devkit/core/src/json/schema/registry.ts#L360 however thats only from a brief look.

**Comment by Jordan-Hall** [comment]
@Ankit-sagar32  have you tried this way:

```
{
  type: "integer",
  minimum: 1,
  maximum: 1000,
  if: {minimum: 100},
  then: {multipleOf: 100},
  else: {
    if: {minimum: 10},
    then: {multipleOf: 10}
  }
}

```

I know officially documentation  says it should us const. But AJV says this is the way.

**Comment by dgp1130** [comment]
Duplicate of #22242.

**Comment by dgp1130** [comment]
Using #22242 as the canonical issue here around the general feature request of "making schematics options more dynamic". Exposing the if/then syntax of JSON schema is probably too specific to this particular dependency and tricky to reason about. Most likely this would like a TypeScript API to prompt users imperatively.
